app/public/cantusdata/helpers/mei_processing/mei_parser.py

```diff
@@ -0,0 +1,274 @@
+            neume_dict: Neume = {
+                "neume_type": neume_type,
+                "neume_components": neume_components,
+                "intervals": intervals,
+                "contours": contours,
+            }
+            neumes_list.append(neume_dict)
+        syllable_dict: Syllable = {
+            "text": text_dict,
+            "neumes": neumes_list,
+        }
+        return syllable_dict
+
+    def parse_mei(self) -> List[Syllable]:
+        """
+        Parses the MEI file into a list of syllables.
+
+        :return: A list of syllables
+        """
+        syllables = []
+        for syllable in self.mei.iter(f"{self.MEINS}syllable"):
+            syllable_dict = self.parse_syllable(syllable)
+            syllables.append(syllable_dict)
+        return syllables
+
+
+def get_interval_between_neume_components(
+    neume_component_1: NeumeComponent,
+    neume_component_2: NeumeComponent,
+) -> int:
+    """
+    Computes the interval (in semitones) between two
+    pitches. Does this by converting the pitches to MIDI note
+    numbers and then subtracting to get difference in semitones.
+    Note: All B's, whether they would be performed as Bb or B,
+    are treated as B.
+
+    :param neume_component_1: A dictionary containing neume component information
+    :param neume_component_2: A dictionary containing neume component information
+    :return: The interval between the two pitches (in semitones)
+    """
+    pc1 = PITCH_CLASS[neume_component_1["pname"]]
+    pc2 = PITCH_CLASS[neume_component_2["pname"]]
+    # In MIDI note numbers, C0 = 12.
+    pitch_1 = pc1 + (12 * (neume_component_1["octave"] + 1))
+    pitch_2 = pc2 + (12 * (neume_component_2["octave"] + 1))
+    return pitch_2 - pitch_1
+
+
+def get_contour_from_interval(interval: int) -> Literal["u", "d", "s"]:
+    """
+    Computes the contour of an interval.
+
+    :param interval: The size of the interval in semitones
+    :return: The contour of the interval ("u"[p], "d"[own], or "s"[tay])
+    """
+    if interval < 0:
+        return "d"
+    if interval > 0:
+        return "u"
+    return "s"
+
+
+def analyze_neume(neume: List[NeumeComponent]) -> Tuple[str, List[int], List[str]]:
+    """
+    Analyze a neume (a list of neume components) to determine:
+    - Neume type
+    - Neume intervals
+    - Neume contour
+
+    :param neume: A list of neume components (a list of NeumeComponentsType dictionaries)
+    :return: A tuple of information about the neume:
+                - Neume type (str)
+                - Neume intervals (list of ints)
+                - Neume contour (list of "u"[p], "d"[own], or "s"[tay])
+    """
+    intervals: List[int] = [
+        get_interval_between_neume_components(nc1, nc2)
+        for nc1, nc2 in zip(neume[:-1], neume[1:])
+    ]
+    contours: List[str] = [get_contour_from_interval(i) for i in intervals]
+    neume_type: str = NEUME_GROUPS.get("".join(contours), "Compound")
```

Comment:
Yes, if not found in `NEUME_GROUPS` we just call it a "Compound" neume (this is how it was previously, so we might want to change this eventually...something to consider ~ I've made a note [here](https://github.com/DDMAL/cantus/issues/796#issuecomment-1883999704) so that we don't lose that)

If `intervals` is [], `contours` is [], so `"".join(contours)` is `""` and `NEUME_GROUPS.get("")` evaluates to `"Punctum"`. We don't differentiate between puncta and virga here.

Reply:
Ah. Didn't catch the empty string in `NEUME_GROUPS`. Very well.